source/server/admin/prometheus_stats.h

```diff
@@ -35,14 +35,14 @@ class PrometheusStatsFormatter {
   static std::string formattedTags(const std::vector<Stats::Tag>& tags);
 
   /**
-   * Format the given metric name, and prefixed with "envoy_" if it does not have a custom
-   * stat namespace. If it has a custom stat namespace AND the name without the custom namespace
-   * has a valid prometheus namespace, the trimmed name is returned.
-   * Otherwise, return nullopt.
+   * Format the given metric name, and prefixed with "envoy_" if custom_stat_namespace_empty=true or
+   * the metric name does not have a custom stat namespace. If it has a custom stat namespace AND
+   * the name without the custom namespace has a valid prometheus namespace, the trimmed name is
+   * returned. Otherwise, return nullopt.
    */
   static absl::optional<std::string>
-  metricName(const std::string& extracted_name,
-             const Stats::CustomStatNamespaces& custom_namespace_factory);
+  metricName(const std::string& extracted_name, const bool custom_stat_namespace_empty,
```

Comment:
it seems awkward to have to pass `foo.empty()` and `foo` into metricName. Maybe you are doing this only to save a virtual function call?

I think it would be cleaner to just change stripRegisteredPrefix to early-exit if empty, which can be discovered locally. I should've suggested that specially earlier. I just wanted to avoid the string-search to find the prefixes, only to then make a lookup into an empty map.

Reply:
thanks, now the code becomes much simpler!